### PR TITLE
sql, bulkio: Support file upload using COPY

### PR DESCRIPTION
**Pull request by g3orgia**
We are currently unable to upload a bulk file to a node's
local file system using a SQL connectiong. This feature
will allow users to upload files so that they can take
advantage of our `IMPORT`'s support for nodelocal, all
through a SQL connection.

We are giving `COPY` the capability to upload when copying
to a specific table. We also expect a destination filename
to be passed in using options. The copying functionality
piggybacks off the current `copyMachine` and changes the path
of execution based on the table name. The unit test `TestFileUpload`
imitates what the client would do to use this feature. This
functionality will later be added to the cockroach CLI.

In order to fit the destination filename into the syntax of
`COPY`, we had to manually build the query on the client side,
as well as change the parsing of COPY queries on the server
side. The `COPY` now accepts statements like:

`COPY table FROM stdin WITH destination = "destination/file"`

We initially intended for this added feature of COPY to be
free of grammar changes. However, there was no good way of
specifying the file destination of the uploaded file. One
potential alternative to this change was to specify this as
the column name of the table to which we are uploading. This
causes concerns with the handling of special characters as well
as placeholders support. We therefore decided that the syntax
change might be worthwhile.

Within this PR, there are no user facing changes (users should
not be running this query manually) and therefore should not be
added to our documentation.

Release note: None

### CODE REVIEW

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42748)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @g3orgia, and @mjibson)

___
*[pkg/sql/copy.go, line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42748#-LuZfOg048AZqUadSc0p:-LuZfOg048AZqUadSc0q:bksvnb1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3625843e79ffe5bc3f9eda827461d4f3504bfc/pkg/sql/copy.go#L98)):*
> ```Go
> 
> 	processRows func(ctx context.Context) error
> 	writeToFile *os.File
> ```

nit: I'd make this a `io.WriteCloser` 

actually, I'd go one step further: I'd make a `type fileUploadMachine struct { copyMachine; sink io.WriteCloser }` that _wraps_ copyMachine so that the one used by SQL doesn't even contain the file code. then `writeFile` is a method only on our wrapper, and 
___
*[pkg/sql/copy.go, line 171 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42748#-LuZkce81tHYl2ubuov8:-LuZkce81tHYl2ubuov9:bpqhy6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3625843e79ffe5bc3f9eda827461d4f3504bfc/pkg/sql/copy.go#L171)):*
> ```Go
> 	}
> 
> 	destination := strings.Trim(n.Options[0].Value.String(), `'"`)
> ```

this looks wrong to me. I think we need to a) call something to correctly load and parse the options (so we should make sure the tests include an example with a bound argument as well as a literal) but we shouldn't , I think, need to have manual string cleaning here.
___
*[pkg/sql/copy.go, line 172 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42748#-LuZleoABo0Yvd9beTM5:-LuZleoABo0Yvd9beTM6:bt7xvq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3625843e79ffe5bc3f9eda827461d4f3504bfc/pkg/sql/copy.go#L172)):*
> ```Go
> 
> 	destination := strings.Trim(n.Options[0].Value.String(), `'"`)
> 	filename := filepath.Join(c.p.execCfg.Settings.ExternalIODir, destination)
> ```

I think this is missing logic for when nodelocal file access is disabled intentionally (which I think is empty string, so right now it'd just write to CWD?) and critically is missing checks for escaping from the IO dir prefix.

I wonder if we should just delegate all of this to MakeExternalStorage? i.e. should we just say the `destination` is an external storage URI? that would remove a bunch of duplicate logic here, and as a bonus (?) could allow uploading _to a specific node_ even if your SQL conn is going though a load-balancer. I guess that raises some _interesting_ questions about if we'd want to allow uploading to anything other than `nodelocal` -- i.e. is there any reason a user should be allowed to upload `WITH destination = 's3://...` ? If we do delegate to MakeExternalStorage, I say we dodge that question for now and just only allow `nodelocal` initially.
___
*[pkg/sql/copy.go, line 265 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42748#-LuZfk5l3-ZbLDWUV_db:-LuZfk5l3-ZbLDWUV_dc:b-1kadsw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3625843e79ffe5bc3f9eda827461d4f3504bfc/pkg/sql/copy.go#L265)):*
> ```Go
> 	// Finalize execution by sending the statement tag and number of rows
> 	// inserted.
> 	c.writeToFile.Close()
> ```

huh, doesn't this need to be wrapped in a non-nil check?

also, I think this needs to be deferred, since there are early returns above, so we want to make sure we don't leak an open file if we never get here. In fact, we probably need to either wait to open the file until the top of this function so we can be _sure_ there is a deferred `Close()`, or we need to add an exported `Close` method to the machine itself and insist that any caller who creates an upload machine must also close it.

**Review by g3orgia — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42748)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @mjibson)

___
*[pkg/sql/copy.go, line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42748#-LuZfOg048AZqUadSc0p:-LucR3FaENi0qXEh9YuX:bndc31p) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3625843e79ffe5bc3f9eda827461d4f3504bfc/pkg/sql/copy.go#L98)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

nit: I'd make this a `io.WriteCloser` 

actually, I'd go one step further: I'd make a `type fileUploadMachine struct { copyMachine; sink io.WriteCloser }` that _wraps_ copyMachine so that the one used by SQL doesn't even contain the file code. then `writeFile` is a method only on our wrapper, and 
</blockquote></details>

That sounds good. This will also make it easier to separate the logic into a new go file, which would make both files much cleaner.
___
*[pkg/sql/copy.go, line 171 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42748#-LuZkce81tHYl2ubuov8:-LucRYP_CW2maY1tXlNn:b-1v7fv6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3625843e79ffe5bc3f9eda827461d4f3504bfc/pkg/sql/copy.go#L171)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

this looks wrong to me. I think we need to a) call something to correctly load and parse the options (so we should make sure the tests include an example with a bound argument as well as a literal) but we shouldn't , I think, need to have manual string cleaning here.
</blockquote></details>

Okay, will investigate further and add the appropriate tests for this.
___
*[pkg/sql/copy.go, line 172 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42748#-LuZleoABo0Yvd9beTM5:-LucRfEs9I5IiWvrpjhn:b-7z9nxq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3625843e79ffe5bc3f9eda827461d4f3504bfc/pkg/sql/copy.go#L172)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I think this is missing logic for when nodelocal file access is disabled intentionally (which I think is empty string, so right now it'd just write to CWD?) and critically is missing checks for escaping from the IO dir prefix.

I wonder if we should just delegate all of this to MakeExternalStorage? i.e. should we just say the `destination` is an external storage URI? that would remove a bunch of duplicate logic here, and as a bonus (?) could allow uploading _to a specific node_ even if your SQL conn is going though a load-balancer. I guess that raises some _interesting_ questions about if we'd want to allow uploading to anything other than `nodelocal` -- i.e. is there any reason a user should be allowed to upload `WITH destination = 's3://...` ? If we do delegate to MakeExternalStorage, I say we dodge that question for now and just only allow `nodelocal` initially.
</blockquote></details>

Hmmm so I've thought about doing this but I believe ExternalStorage has other properties we may not need. I will double check this, but something else we could potentially use is `localStorage`, which only deals with the node's local file system. It is currently private within the blob service, but we could expose it, or move it to a utils package.
___
*[pkg/sql/copy.go, line 265 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42748#-LuZfk5l3-ZbLDWUV_db:-LucSRps4vLlGbyycwMy:b6oqbr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3625843e79ffe5bc3f9eda827461d4f3504bfc/pkg/sql/copy.go#L265)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

huh, doesn't this need to be wrapped in a non-nil check?

also, I think this needs to be deferred, since there are early returns above, so we want to make sure we don't leak an open file if we never get here. In fact, we probably need to either wait to open the file until the top of this function so we can be _sure_ there is a deferred `Close()`, or we need to add an exported `Close` method to the machine itself and insist that any caller who creates an upload machine must also close it.
</blockquote></details>

Ah yes, this definitely requires refining. Sorry I missed this, it was only meant to be in the prototype. I think giving the `copyMachine` or `fileUploadMachine` a close method would probably be the best way to go here.

**Review by knz — COMMENTED** [review]
> Within this PR, there are no user facing changes (users should not be running this query manually).

This is a flawed interpretation of 'no user facing change'. Any change to the SQL grammar is a user-facing change. More so that we are in the process of documenting every aspect of `crdb_internal` so this change would land in docs at that point.

That being said, I wonder if `COPY crdb_internal.file_upload FROM STDIN WITH destination = <file>` is the right syntax to go about this? It sounds overly verbose, and arguably an abuse of the `COPY <destination_table>` syntax. 

Why did you not choose to go for a dedicated syntax using simply `COPY '<file>' FROM STDIN` ? i.e. using a SCONST directly after COPY. If you look at the grammar, SCONST is not allowed directly after COPY so it would not be ambiguous.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42748#-:-LulFvTz7dSMc_jJu1Gd:b-9txtc5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @mjibson)

**Comment by dt** [comment]
The special-case table name of `crdb_internal.file_upload` was my suggestion, originally with the intent to avoid _any_ grammar changes at all: I envisioned reusing the existing `COPY table(columns)` syntax as-is, by using a special meta table as the target, that when “written to”, write to a file instead, with the “target column”  of said table identifying the file name, thus fitting within the confines of the existing syntax. I think the move to use the additional options syntax here though was out of concerns about what special characters or escaping might be required in a column name and placeholder support.

Note that there is a strong concern around not inventing a completely new syntax as drivers (including libpq) sometimes look for the literal syntax when deciding to switch into the copy format/protocol, so we specifically wanted to use the original syntax so that drivers would work. Libpq even wants to write the COPY _statement_ for you via a (pure) function that takes the table and column names, likely so it can later then be sure of its form and do a string prefix comparison against the exact prefix `COPY ` . 

Anyway, that’s background on why this prototype went with this syntax and meta table name, though at least with libpq I _think_ we might be able to get away with the SCONST if it is only looking at the first 5 bytes?

As far as the user-facing part of this, I’d prefer to simply hide the new syntax, whichever it is, from the docs than call this a a documented, public (and this stable) public API. This is _internal_ syntax for use exclusively by the CLI, at least for now which this is an alpha feature subject to change.

**Comment by knz** [comment]
Thanks for explaining. The desire to reuse existing copy APIs in drivers is compelling.

It would definitely help the reviewers and later maintainers to include the entirety of this rationale in the commit message and PR description.

Moreover I would love to see a confirming copy test using lib/pq as part of the PR, is it included? (I might have missed it - I did not review the code in detail since I am not primary reviewer).

Besides this, there is a special annotation that can be used in the grammar to avoid including the extension in the generated syntax diagrams. Or alternatively modify the includes/excludes in cmd/docgen.
-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.

**Comment by knz** [comment]
I also would encourage to try harder and use the special column name idea.

I'll be playing devil's advocate here: if you find that the base COPY syntax is insufficient for your use cases of interest, that would cancel your argument that the current approach is drop in compatible with existing copy APIs (you would need to bypass said standard copy APIs to pass your new "with" options).

The ideas to use a special table name and alsp custom kvoptions are contradictory IMHO.
-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.

**Comment by dt** [comment]
I believe the special column name was later rejected, since column names cannot support placeholders and placeholders seemed like a must-have for scary things in file names. The trailing position of the `WITH ` options I believe means that they shouldn't interfere with, as far as we know, the statement fingerprinting done in the drivers to determine if they should switch modes. That said, the only driver we care about _initially_ is the one our shell uses -- libpq. The current tests here are exercising it I think -- the connection returned by `StartServer` is backed by go db/sql.Open("postgres", <server URI>) which, as far as I know, is a libpq connection?

**Comment by knz** [comment]
FYI We're transitioning to pgx to replace lib/pq in the 20.1 timeframe. You'll want to check that your solution works for pgx too.
-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.

**Comment by knz** [comment]
Also FYI I went to the `lib/pq` source code and the detection is done 

1) only for prepared statements

2) on just the 4 first characters (not even including the space afterwards)

Additionally it has a "copy query constructor" (`func CopyIn(...) string`)  which only uses the basic form but does not add a final semicolon, so you can tack on an extra `with` clause at the end, before you pass the produced string as input to `Prepare()`.

I don't know whether `lib/pq` supports non-prepared COPY and it does not seem to be the case based on my (casual) analysis of the source.

Meanwhile, `pgx` does not work this way. It has two separate APIs for COPY:

1. an API for the pgwire binary protocol called `CopyFrom()`, which takes a table name and column set as destination argument, and an iterator as data input. This does not let you control the SQL used *at all* - it's always "`copy <table> (<cols>) from stdin binary;`"

2. an API for the pgwire text protocol called `CopyFromReader()`, which takes as input the SQL of a COPY statement as a string (the client is responsible for creating it) and a Go `io.Reader` to provide the input data.

What this analysis of lib/pq and pgx tells us:

- any syntax extension to COPY (including a `with` clause) is going to give us trouble with *some* pre-defined "native" driver support, such as the `CopyFrom` API in `pgx` - but not all

- at least in these two drivers, there is at least one way to use a custom COPY syntax (such as `COPY <filename as SCONST>`)

tldr - either:

- you fit all the semantics in the base pg syntax, in which case you enable *all* the native driver APIs; or
- you can't, in which case you have to give up *some* native driver support anyway, but there is enough remaining to support _any_ custom syntax that starts with the COPY keyword.

**Comment by g3orgia** [comment]
@knz @dt Thank you for bringing up these concerns. I've updated the commit message and description of this PR to explain the design decisions that were made. I was unaware that we are moving away from `lib/pq` to `pgx` and can see why these changes would cause concerns.

I noticed that the latest version of `pgx` seem to have removed the `CopyFromReader` API for copy. Although we are not currently using that version, I'm hesitant to use that API for these changes right now. I am currently unsure how to move forward, should I:
- Use the existing design, and `lib/pq`, if we will still be using it in 20.1. Will we be completely moving off `lib/pq`?
- Use the existing design with `pgx`'s `CopyFromReader`. However, if we upgrade to latest version then it would no longer exist, and we'd have to find a workaround.
-  Revert the syntax changes and use the column name instead. This would probably require testing the edge cases of supported characters in column names, and limiting destination file name to those characters from the CLI. But it will be compatible with all driver APIs for copy.

**Comment by knz** [comment]
We can also approach the pgx authors with our concerns and request something that works for us.
-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.

**Comment by mjibson** [comment]
pgx has refactored itself and it now has a `pgconn` package with a [`CopyFrom`](https://github.com/jackc/pgconn/blob/32350bd1dc3288aa22f271f87065741da0a1bdb8/pgconn.go#L949) method that I think will allow us to do whatever we need since it takes a io.Reader and sql string.

File: pkg/sql/sem/tree/copy.go
Comment (mjibson) [inline pkg/sql/sem/tree/copy.go]: You need to add one of these to sql/parser/parse_test.go which will verify that we can parse, print, and pretty print these correctly.

File: pkg/sql/conn_executor.go
Comment (mjibson) [inline pkg/sql/conn_executor.go]: Can you refactor this func out into a var so it can be reused between both if blocks?

File: pkg/sql/copy.go
Comment (mjibson) [inline pkg/sql/copy.go]: `c.processRows = c.insertRows`

File: pkg/sql/copy_file_upload.go
Comment (mjibson) [inline pkg/sql/copy_file_upload.go]: `c.processRows = f.writeFile`

File: pkg/sql/conn_executor.go
Comment (mjibson) [inline pkg/sql/conn_executor.go]: Should we verify that cmd.Stmt.Columns is empty? It should probably be an error if it's not.

File: pkg/sql/copy_file_upload.go
Comment (mjibson) [inline pkg/sql/copy_file_upload.go]: For the file uploader, this and columns can be unset because they aren't needed, so we shouldn't set them. If they are empty and required, it might cause a loud error, which is good. (There's some refactoring work that could be done here to remove the dependency on copyMachine.tables/columns from insertRows where it's currently used, but only do that if you're feeling ambitious.)

File: pkg/sql/copy_file_upload.go
Comment (mjibson) [inline pkg/sql/copy_file_upload.go]: why does this type need to exist?

File: pkg/sql/copy_file_upload.go
Comment (mjibson) [inline pkg/sql/copy_file_upload.go]: can this just be `chan error`?

File: pkg/sql/copy_file_upload.go
Comment (mjibson) [inline pkg/sql/copy_file_upload.go]: this needs to be escaped since it could contain, say, a single quote.

File: pkg/sql/copy_file_upload.go
Comment (mjibson) [inline pkg/sql/copy_file_upload.go]: This chan needs to be size 1 otherwise it could leak goroutines. The happy case below is that localStorage.WriteFile completes successfully and attempts to send it over the chan. Once something receives over the chan, it closes the channel and exits the goroutine. Yay.

In fileUploadMachine.run, a chan receive is only done at the very end of the function. But there are two other returns. If either of those is taken, then this goroutine will block forever waiting to send on the chan, but the only possible receiver already exited.

If you instead make this chan buffered with length 1, then this goroutine is guaranteed to always be able to send on the chan, which will allow for both the go routine and the chan to be cleaned up in either success or error conditions.

**Comment by g3orgia** [comment]
Hmmm yes that is the case, so if we were to migrate to the new version of `pgx`, we could definitely use the underlying `CopyFrom` directly (which actually provides the perfect API). I will document as future work, and keep working on the CLI using the `lib/pq` implementation. In this case, this PR is ready for review. :)

File: pkg/sql/conn_executor.go
Comment (g3orgia) [reply pkg/sql/conn_executor.go]: We're breaking up the file content into rows of one column, so it shouldn't be empty as of right now. I can check that there is exactly one row though.

File: pkg/sql/copy_file_upload.go
Comment (g3orgia) [reply pkg/sql/copy_file_upload.go]: Okay, will add a TODO to remove this dependency from insertRows.

File: pkg/sql/copy_file_upload.go
Comment (g3orgia) [reply pkg/sql/copy_file_upload.go]: No, it doesn't I had some confusion about the behaviour of goroutines, removing this.

File: pkg/sql/copy_file_upload.go
Comment (g3orgia) [reply pkg/sql/copy_file_upload.go]: Thank you for clearing this up for me! I'm new to channels and found it a little hard to understand and research. This mentioned some keywords that helped a lot, along with the behaviors in different cases. Making those changes!

File: pkg/sql/copy_file_upload.go
Comment (g3orgia) [reply pkg/sql/copy_file_upload.go]: No, it doesn't I had some confusion about the behaviour of goroutines, removing this.

File: pkg/sql/conn_executor.go
Comment (mjibson) [reply pkg/sql/conn_executor.go]: I'm not sure we're talking about the same thing. I'm talking about the tree.CopyFrom node which could have columns. Like we should ensure a user doesn't do `COPY crdb_internal.file_upload(blah) TO ...` since the `blah` is completely ignored and may possibly do something the user doesn't expect.

File: pkg/sql/copy_file_upload_test.go
Comment (mjibson) [inline pkg/sql/copy_file_upload_test.go]: why does this table need to exist?

**Review by madelynnblue — APPROVED** [review]
LGTM, assuming everyone is happy with the current design.

File: pkg/sql/conn_executor.go
Comment (g3orgia) [reply pkg/sql/conn_executor.go]: Ah I see, that's fair. I think currently, I actually do pass in a column name. That could cause some confusion since it's unused, will change it so that there is no column name, and check that it doesn't have it.

File: pkg/sql/copy_file_upload_test.go
Comment (g3orgia) [reply pkg/sql/copy_file_upload_test.go]: Oof, good catch this was from a previous version of the prototype, when I was trying things out. Will remove it.

**Review by madelynnblue — APPROVED** [review]

**Comment by g3orgia** [comment]
@dt lemme know if you'd like another pass at this PR, or if it's ready to bors. :)

**Comment by bdarnell** [comment]
What permissions does this require? I have several concerns including a low-privilege user filling up the server's disk, and a low-privilege user either overwriting a file that exists, or squatting on a filename that another user intends to use. I think it's probably best to make this admin-only for now.

**Comment by dt** [comment]
I'm fine with making file uploads Admin only (or a grantable privilege though given our current model of grants living on dbs and tables I don't know how we'd do that without introducing some dummy table). "owning" uploaded files is an explicit non-goal for now (we're not taking on any management, like TTLs or whatever, so we also don't have anything tracking who can read or delete which file, so Admin-only solves that too).

**Pull request merged**
